overlord/devicestate/firstboot.go

```diff
@@ -129,93 +129,107 @@ func populateStateFromSeedImpl(st *state.State, tm timings.Measurer) ([]*state.T
 
 	tsAll := []*state.TaskSet{}
 	configTss := []*state.TaskSet{}
+	chainTs := func(all []*state.TaskSet, ts *state.TaskSet) []*state.TaskSet {
+		n := len(all)
+		if n != 0 {
+			ts.WaitAll(all[n-1])
+		}
+		return append(all, ts)
+	}
 
 	baseSnap := "core"
+	classicWithSnapd := false
 	if model.Base() != "" {
 		baseSnap = model.Base()
 	}
+	if _, ok := seeding["snapd"]; release.OnClassic && ok {
+		classicWithSnapd = true
+		// there is no system-wide base as such
+		// if there is a gadget we will install its base first though
+		baseSnap = ""
+	}
 
-	installSeedEssential := func(snapName string, last int) (*snap.Info, error) {
+	var installSeedEssential func(snapName string) error
+	var installGadgetBase func(gadget *snap.Info) error
+	installSeedEssential = func(snapName string) error {
+		// be idempotent for installGadgetBase use
+		if alreadySeeded[snapName] {
+			return nil
+		}
 		seedSnap := seeding[snapName]
 		if seedSnap == nil {
-			return nil, fmt.Errorf("cannot proceed without seeding %q", snapName)
+			return fmt.Errorf("cannot proceed without seeding %q", snapName)
 		}
 		ts, info, err := installSeedSnap(st, seedSnap, snapstate.Flags{SkipConfigure: true, Required: true}, tm)
 		if err != nil {
-			return nil, err
+			return err
 		}
-		if last >= 0 {
-			ts.WaitAll(tsAll[last])
+		if info.GetType() == snap.TypeGadget {
```

Comment:
Sill question: could this block go before `ts, info, err := installSeedSnap(...` above to reflect installation order? If I'm reading this correctly, we have a tasket for gadget itself, then "recursively" get into `installSeedEssential` again to install gadget's base and the taskset chaining is correct thanks to delayed chaining below, which is a bit confusing. IMHO it would be clearer if `ts` was immediately added to the chain after `installSeedSnap`, and the operations were reordered; or am I missing somthing?

Reply:
we need info to make that decision, so not sure how it can go before?  also notice that in the end we sort the ts before actually chaining them together so the order is not relevant

Reply:
Ah, of course, it needs the infom, scratch that. Thanks.